File: vihChatBot/src/main/java/com/vihmessenger/vihchatbot/data/services/ApiService.kt
package com.vihmessenger.vihchatbot.data.services

import com.vihmessenger.vihchatbot.constants.BaseAPIConstants
import com.vihmessenger.vihchatbot.data.model.ChatHistoryModel
import com.vihmessenger.vihchatbot.data.model.ChatListModelResponse
import com.vihmessenger.vihchatbot.data.model.ChatMessageModel
import com.vihmessenger.vihchatbot.data.model.DeviceTokenRequest
import com.vihmessenger.vihchatbot.data.model.DeviceTokenResponse
import com.vihmessenger.vihchatbot.data.model.EnterPriseDiscoverModel
import com.vihmessenger.vihchatbot.data.model.EnterpriseApiResponse
import com.vihmessenger.vihchatbot.data.model.CallDetailsRequest
import com.vihmessenger.vihchatbot.data.model.IndustryResponse
import com.vihmessenger.vihchatbot.data.model.EmailLoginRequest
import com.vihmessenger.vihchatbot.data.model.EmailLoginResponse
import com.vihmessenger.vihchatbot.data.model.SubscribeChannelRequest
import com.vihmessenger.vihchatbot.data.model.SubscribeChannelResponse
import com.vihmessenger.vihchatbot.data.model.LoanApprovalRequest
import com.vihmessenger.vihchatbot.data.model.LoanApprovalResponse
import com.vihmessenger.vihchatbot.data.model.LogoutDataModel
import com.vihmessenger.vihchatbot.data.model.SdkFeatureResponse
import com.vihmessenger.vihchatbot.data.model.UpdateUserProfile
import com.vihmessenger.vihchatbot.data.model.UserProfilePatchUsername
import com.vihmessenger.vihchatbot.data.model.UserProfileRequest
import com.vihmessenger.vihchatbot.data.model.UserProfileResponse
import com.vihmessenger.vihchatbot.data.model.UserProfileUpdateResponse
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.Multipart
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Part
import retrofit2.http.PartMap
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Url
import com.google.gson.JsonObject

/**
 * SECURITY: Authorization header is now injected centrally via AuthInterceptor.
 * Removed inline "Bearer ${AppController.prefs?.accessToken}" default parameters
 * which were vulnerable to race conditions and could send "Bearer null".
 */
interface ApiService : BaseApiService {

    // Trailing slash required by the DRF SimpleRouter — without it the server 301-redirects.
    @GET(BaseAPIConstants.SDK_FEATURES + "{hashCode}/")
    suspend fun getSdkFeatures(@Path("hashCode") hashCode: String): Response<SdkFeatureResponse>

    @POST(BaseAPIConstants.USER_SIGNUP_LOGIN)
    suspend fun createUserProfile(@Body body: UserProfileRequest): Response<UserProfileResponse>

    // Email-OTP token exchange. Body carries the verified Cognito ID token (auth) plus
    // mobile/channel_id for hashcode-matched delivery; returns the existing app-session
    // tokens or account_status == "needs_profile".
    @POST(BaseAPIConstants.EMAIL_LOGIN)
    suspend fun emailLogin(@Body body: EmailLoginRequest): Response<EmailLoginResponse>

    // Subscribe the authenticated user to a channel (Settings hashkey switch).
    @POST(BaseAPIConstants.SUBSCRIBE_CHANNEL)
    suspend fun subscribeChannel(@Body body: SubscribeChannelRequest): Response<SubscribeChannelResponse>

    @PATCH(BaseAPIConstants.USER_PROFILE)
    suspend fun updateUserProfile(
        @Body body: UpdateUserProfile
    ): Response<UserProfileUpdateResponse>

    @PATCH(BaseAPIConstants.USER_PROFILE)
    suspend fun updateUserProfileUsername(
        @Body body: UserProfilePatchUsername
    ): Response<UserProfileUpdateResponse>

    @Multipart
    @PATCH(BaseAPIConstants.USER_PROFILE)
    suspend fun updateUserProfileImage(
        @Part image: MultipartBody.Part
    ): Response<UserProfileUpdateResponse>

    @Multipart
    @PATCH(BaseAPIConstants.USER_PROFILE)
    suspend fun updateProfileSelective(
        @PartMap partMap: HashMap<String, RequestBody>,
        @Part user_profile_image: MultipartBody.Part?
    ): Response<UserProfileUpdateResponse>

    @GET(BaseAPIConstants.MAIN_CHAT + "{channel_id}" + "/" + "{enterprise_id}")
    suspend fun getChatResponse(
        @Path("channel_id") hashcode: String,
        @Path("enterprise_id") enterpriseId: String,
        @Query("question") question: String,
        @Query("session_id") session_id: String
    ): Response<ChatMessageModel>

    @GET(BaseAPIConstants.CHAT_HISTORY + "{channel_id}" + "/" + "{enterprise_id}")
    suspend fun getChatHistory(
        @Path("channel_id") channel_id: String,
        @Path("enterprise_id") enterprise_id: String
    ): Response<ChatHistoryModel>

    @GET(BaseAPIConstants.MAIN_CHAT_LIST)
    suspend fun getChatListResponse(
        @Query("channel_id") hashcode: String,
        @Header("Content-Type") contentTypeUrlEncoded: String = "application/x-www-form-urlencoded"
    ): Response<ChatListModelResponse>

    @GET(BaseAPIConstants.MAIN_DISCOVER_LIST)
    suspend fun getEnterpriseDiscoverListResponse(
        @Query("channel_id") hashcode: String,
        @Query("page") page: Int,
        @Query("search") search: String,
        @Query("industries") industries: String
    ): Response<EnterPriseDiscoverModel>

    @GET(BaseAPIConstants.INDUSTRIES)
    suspend fun getIndustries(): Response<IndustryResponse>

    @GET(BaseAPIConstants.ENTERPRISES)
    suspend fun getEnterprises(@Query("enterprise_id") enterprise_id: String): Response<EnterpriseApiResponse>

    @Multipart
    @PATCH(BaseAPIConstants.USER_PROFILE)
    suspend fun createProfile(
        @PartMap partMap: MutableMap<String, RequestBody>,
        @Part user_profile_image: MultipartBody.Part
    ): Response<UserProfileUpdateResponse>

    @FormUrlEncoded
    @POST(BaseAPIConstants.USER_LOGOUT)
    suspend fun userLogout(
        @Field("refresh_token") refresh_token: String
    ): Response<LogoutDataModel>

    @POST(BaseAPIConstants.REGISTER_FCM_TOKEN)
    suspend fun registerFcmToken(
        @Body body: DeviceTokenRequest
    ): Response<DeviceTokenResponse>

    @POST
    suspend fun postLoanApproval(
        @Url url: String,
        @Body body: LoanApprovalRequest
    ): Response<LoanApprovalResponse>

    @POST
    suspend fun postCallDetails(
        @Url url: String,
        @Body body: CallDetailsRequest
    ): Response<JsonObject>
}

File: Sources/VihChatBotSDK/Networking/ApiService.swift
import Foundation

/// Swift equivalent of Retrofit's `ApiService` interface. Each `suspend fun`
/// becomes an `async throws` method. Path/query/body assembly that Retrofit
/// handled via annotations is done inline below.
///
/// All Authorization header injection runs through `AuthInterceptor` exactly
/// like Android — never set it inline (matches the security note in `ApiService.kt`).
public protocol ApiService {
    func getSdkFeatures(hashCode: String) async throws -> SdkFeatureResponse

    func createUserProfile(body: UserProfileRequest) async throws -> UserProfileResponse

    func emailLogin(body: EmailLoginRequest) async throws -> EmailLoginResponse

    func subscribeChannel(body: SubscribeChannelRequest) async throws -> SubscribeChannelResponse

    func updateUserProfile(body: UpdateUserProfile) async throws -> UserProfileUpdateResponse

    func updateUserProfileUsername(body: UserProfilePatchUsername) async throws -> UserProfileUpdateResponse

    func updateUserProfileImage(imageData: Data, mimeType: String, filename: String) async throws -> UserProfileUpdateResponse

    func updateProfileSelective(
        fields: [String: String],
        imageData: Data?,
        imageMimeType: String?,
        imageFilename: String?
    ) async throws -> UserProfileUpdateResponse

    func getChatResponse(
        hashcode: String,
        enterpriseId: String,
        question: String,
        sessionId: String
    ) async throws -> ChatMessageModel

    func getChatHistory(channelId: String, enterpriseId: String) async throws -> ChatHistoryModel

    func getChatListResponse(hashcode: String) async throws -> ChatListModelResponse

    func getEnterpriseDiscoverList(
        hashcode: String,
        page: Int,
        search: String,
        industries: String
    ) async throws -> EnterPriseDiscoverModel

    func getIndustries() async throws -> IndustryResponse

    func getEnterprises(enterpriseId: String) async throws -> EnterpriseApiResponse

    func createProfile(
        fields: [String: String],
        imageData: Data,
        imageMimeType: String,
        imageFilename: String
    ) async throws -> UserProfileUpdateResponse

    func userLogout(refreshToken: String) async throws -> LogoutDataModel

    func registerDeviceToken(body: DeviceTokenRequest) async throws -> DeviceTokenResponse

    // Per-enterprise state mutations (see EnterPriseModel flags).
    func blacklistEnterprise(enterprisePk: Int, blacklist: Bool) async throws -> GenericStatusResponse
    func muteEnterprise(enterpriseId: Int, muteStatus: Bool) async throws -> GenericStatusResponse
    func updateEnterprisePromotional(optIn: Bool, enterpriseId: Int, channelId: String) async throws -> GenericStatusResponse
}

final class APIServiceImpl: ApiService {

    private let client: APIClient
    init(client: APIClient) { self.client = client }

    // MARK: - SDK features / chat / enterprises

    func getSdkFeatures(hashCode: String) async throws -> SdkFeatureResponse {
        // Trailing slash required by the DRF SimpleRouter — without it the server 301-redirects.
        let req = client.request(path: BaseAPIConstants.sdkFeatures + hashCode + "/")
        return try await client.send(req, as: SdkFeatureResponse.self)
    }

    func createUserProfile(body: UserProfileRequest) async throws -> UserProfileResponse {
        let req = client.request(
            path: BaseAPIConstants.userSignupLogin,
            method: "POST",
            body: try JSONEncoder().encode(body),
            contentType: "application/json"
        )
        return try await client.send(req, as: UserProfileResponse.self)
    }

    func emailLogin(body: EmailLoginRequest) async throws -> EmailLoginResponse {
        let req = client.request(
            path: BaseAPIConstants.emailLogin,
            method: "POST",
            body: try JSONEncoder().encode(body),
            contentType: "application/json"
        )
        return try await client.send(req, as: EmailLoginResponse.self)
    }

    func subscribeChannel(body: SubscribeChannelRequest) async throws -> SubscribeChannelResponse {
        let req = client.request(
            path: BaseAPIConstants.subscribeChannel,
            method: "POST",
            body: try JSONEncoder().encode(body),
            contentType: "application/json"
        )
        return try await client.send(req, as: SubscribeChannelResponse.self)
    }

    func updateUserProfile(body: UpdateUserProfile) async throws -> UserProfileUpdateResponse {
        let req = client.request(
            path: BaseAPIConstants.userProfile,
            method: "PATCH",
            body: try JSONEncoder().encode(body),
            contentType: "application/json"
        )
        return try await client.send(req, as: UserProfileUpdateResponse.self)
    }

    func updateUserProfileUsername(body: UserProfilePatchUsername) async throws -> UserProfileUpdateResponse {
        let req = client.request(
            path: BaseAPIConstants.userProfile,
            method: "PATCH",
            body: try JSONEncoder().encode(body),
            contentType: "application/json"
        )
        return try await client.send(req, as: UserProfileUpdateResponse.self)
    }

    func updateUserProfileImage(imageData: Data, mimeType: String, filename: String) async throws -> UserProfileUpdateResponse {
        let multipart = MultipartFormData()
        multipart.appendFile(name: "image", data: imageData, mimeType: mimeType, filename: filename)
        let req = client.request(
            path: BaseAPIConstants.userProfile,
            method: "PATCH",
            body: multipart.bodyData,
            contentType: multipart.contentType
        )
        return try await client.send(req, as: UserProfileUpdateResponse.self)
    }

    func updateProfileSelective(
        fields: [String: String],
        imageData: Data?,
        imageMimeType: String?,
        imageFilename: String?
    ) async throws -> UserProfileUpdateResponse {
        let multipart = MultipartFormData()
        for (k, v) in fields { multipart.appendField(name: k, value: v) }
        if let data = imageData, let mime = imageMimeType, let name = imageFilename {
            multipart.appendFile(name: "user_profile_image", data: data, mimeType: mime, filename: name)
        }
        let req = client.request(
            path: BaseAPIConstants.userProfile,
            method: "PATCH",
            body: multipart.bodyData,
            contentType: multipart.contentType
        )
        return try await client.send(req, as: UserProfileUpdateResponse.self)
    }

    func getChatResponse(
        hashcode: String,
        enterpriseId: String,
        question: String,
        sessionId: String
    ) async throws -> ChatMessageModel {
        // Trailing slash avoids Django's APPEND_SLASH 301 redirect (which iOS
        // would follow without the auth header). Matches the slash-terminated
        // convention of the list endpoints.
        let path = "\(BaseAPIConstants.mainChat)\(hashcode)/\(enterpriseId)/"
        let query = [
            URLQueryItem(name: "question", value: question),
            URLQueryItem(name: "session_id", value: sessionId)
        ]
        let req = client.request(path: path, query: query)
        return try await client.send(req, as: ChatMessageModel.self)
    }

    func getChatHistory(channelId: String, enterpriseId: String) async throws -> ChatHistoryModel {
        let path = "\(BaseAPIConstants.chatHistory)\(channelId)/\(enterpriseId)/"
        let req = client.request(path: path)
        return try await client.send(req, as: ChatHistoryModel.self)
    }

    func getChatListResponse(hashcode: String) async throws -> ChatListModelResponse {
        let req = client.request(
            path: BaseAPIConstants.mainChatList,
            query: [URLQueryItem(name: "channel_id", value: hashcode)],
            headers: ["Content-Type": "application/x-www-form-urlencoded"]
        )
        return try await client.send(req, as: ChatListModelResponse.self)
    }

    func getEnterpriseDiscoverList(
        hashcode: String,
        page: Int,
        search: String,
        industries: String
    ) async throws -> EnterPriseDiscoverModel {
        let query = [
            URLQueryItem(name: "channel_id", value: hashcode),
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "search", value: search),
            URLQueryItem(name: "industries", value: industries)
        ]
        let req = client.request(path: BaseAPIConstants.mainDiscoverList, query: query)
        return try await client.send(req, as: EnterPriseDiscoverModel.self)
    }

    func getIndustries() async throws -> IndustryResponse {
        let req = client.request(path: BaseAPIConstants.industries)
        return try await client.send(req, as: IndustryResponse.self)
    }

    func getEnterprises(enterpriseId: String) async throws -> EnterpriseApiResponse {
        let req = client.request(
            path: BaseAPIConstants.enterprises,
            query: [URLQueryItem(name: "enterprise_id", value: enterpriseId)]
        )
        return try await client.send(req, as: EnterpriseApiResponse.self)
    }

    func createProfile(
        fields: [String: String],
        imageData: Data,
        imageMimeType: String,
        imageFilename: String
    ) async throws -> UserProfileUpdateResponse {
        let multipart = MultipartFormData()
        for (k, v) in fields { multipart.appendField(name: k, value: v) }
        multipart.appendFile(name: "user_profile_image", data: imageData, mimeType: imageMimeType, filename: imageFilename)
        let req = client.request(
            path: BaseAPIConstants.userProfile,
            method: "PATCH",
            body: multipart.bodyData,
            contentType: multipart.contentType
        )
        return try await client.send(req, as: UserProfileUpdateResponse.self)
    }

    func userLogout(refreshToken: String) async throws -> LogoutDataModel {
        // Retrofit @FormUrlEncoded body — Swift equivalent is the same percent-
        // encoded string. Mirrors `apiService.userLogout(refresh_token)`.
        let bodyString = "refresh_token=\(refreshToken.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? refreshToken)"
        let req = client.request(
            path: BaseAPIConstants.userLogout,
            method: "POST",
            body: bodyString.data(using: .utf8),
            contentType: "application/x-www-form-urlencoded"
        )
        return try await client.send(req, as: LogoutDataModel.self)
    }

    func registerDeviceToken(body: DeviceTokenRequest) async throws -> DeviceTokenResponse {
        let req = client.request(
            path: BaseAPIConstants.registerDeviceToken,
            method: "POST",
            body: try JSONEncoder().encode(body),
            contentType: "application/json"
        )
        return try await client.send(req, as: DeviceTokenResponse.self)
    }

    // MARK: - Per-enterprise state mutations

    /// Form-url-encode a flat param map (mirrors Retrofit @FormUrlEncoded @Field). Booleans
    /// are sent as "true"/"false" to match Retrofit's Boolean serialization.
    private func formEncoded(_ params: [String: String]) -> Data? {
        params.map { key, value in
            let k = key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? key
            let v = value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
            return "\(k)=\(v)"
        }.joined(separator: "&").data(using: .utf8)
    }

    func blacklistEnterprise(enterprisePk: Int, blacklist: Bool) async throws -> GenericStatusResponse {
        let req = client.request(
            path: BaseAPIConstants.userBlacklistEnterprise,
            method: "POST",
            body: formEncoded(["enterprise_pk": String(enterprisePk), "blacklist": blacklist ? "true" : "false"]),
            contentType: "application/x-www-form-urlencoded"
        )
        return try await client.send(req, as: GenericStatusResponse.self)
    }

    func muteEnterprise(enterpriseId: Int, muteStatus: Bool) async throws -> GenericStatusResponse {
        let req = client.request(
            path: BaseAPIConstants.muteEnterprise,
            method: "POST",
            body: formEncoded(["enterprise_id": String(enterpriseId), "mute_status": muteStatus ? "true" : "false"]),
            contentType: "application/x-www-form-urlencoded"
        )
        return try await client.send(req, as: GenericStatusResponse.self)
    }

    func updateEnterprisePromotional(optIn: Bool, enterpriseId: Int, channelId: String) async throws -> GenericStatusResponse {
        let req = client.request(
            path: BaseAPIConstants.userChannelEnterpriseConfig,
            method: "POST",
            body: formEncoded([
                "promotional_opt_in": optIn ? "true" : "false",
                "enterprise_id": String(enterpriseId),
                "channel_id": channelId
            ]),
            contentType: "application/x-www-form-urlencoded"
        )
        return try await client.send(req, as: GenericStatusResponse.self)
    }
}
